relay-server/src/services/health_check.rs

```diff
@@ -182,127 +208,61 @@ impl HealthCheckService {
-    }
 }
 
 impl Service for HealthCheckService {
     type Interface = HealthCheck;
 
-    fn spawn_handler(self, mut rx: relay_system::Receiver<Self::Interface>) {
-        let service = Arc::new(self);
+    fn spawn_handler(mut self, mut rx: relay_system::Receiver<Self::Interface>) {
+        let (update_tx, update_rx) = watch::channel(StatusUpdate::healthy());
+        let check_interval = self.config.health_sys_info_refresh_interval();
+        // Add 10% buffer to the internal timeouts to avoid race conditions.
+        let status_timeout = (check_interval + self.config.health_probe_timeout()).mul_f64(1.1);
 
         tokio::spawn(async move {
-            let mut shutdown = Controller::shutdown_handle();
-
-            loop {
-                tokio::select! {
-                    biased;
-
-                    Some(message) = rx.recv() => {
-                        let service = service.clone();
-                        tokio::spawn(async move { service.handle_message(message).await });
-                    }
-                    _ = shutdown.notified() => {
-                        service.is_shutting_down.store(true, Ordering::Relaxed);
-                    }
-                }
-            }
-        });
-    }
-}
+            let shutdown = Controller::shutdown_handle();
 
-#[derive(Debug)]
-struct SystemInfo {
-    system: System,
-    last_refresh: Instant,
-    refresh_interval: Duration,
-}
+            while shutdown.get().is_none() {
+                let _ = update_tx.send(StatusUpdate::new(relay_statsd::metric!(
+                    timer(RelayTimers::HealthCheckDuration),
+                    type = "readiness",
+                    { self.check_readiness().await }
+                )));
 
-impl SystemInfo {
-    /// Creates a new [`SystemInfo`] to query system information.
-    ///
-    /// System information updates are debounced with the passed `refresh_interval`.
-    pub fn new(refresh_interval: Duration) -> Self {
-        let system = System::new_with_specifics(
-            RefreshKind::new().with_memory(MemoryRefreshKind::everything()),
-        );
-
-        Self {
-            system,
-            last_refresh: Instant::now(),
-            refresh_interval,
-        }
-    }
-
-    /// Current snapshot of system memory.
-    ///
-    /// On Linux systems it uses the cgroup limits to determine used and total memory,
-    /// if available.
-    pub fn memory(&mut self) -> Memory {
-        self.refresh();
-
-        // Use the cgroup if available in case Relay is running in a container.
-        if let Some(cgroup) = self.system.cgroup_limits() {
-            Memory {
-                used: cgroup.total_memory.saturating_sub(cgroup.free_memory),
-                total: cgroup.total_memory,
-            }
-        } else {
-            Memory {
-                used: self.system.used_memory(),
-                total: self.system.total_memory(),
+                tokio::time::sleep(check_interval).await;
```

Comment:
Note that in many other places we instead create a `tokio::time::interval` and then use `tokio::select!` with the shutdown handle and the interval. I opted for a different solution here as we don't need to stick to strict intervals and this version ended up less code. The alternative would look like this:

```rust
let mut ticker = tokio::time::interval(check_interval);
let mut shutdown = Controller::shutdown_handle();

loop {
    let _ = update_tx.send(StatusUpdate::new(relay_statsd::metric!(
        timer(RelayTimers::HealthCheckDuration),
        type = "readiness",
        { self.check_readiness().await }
    )));

    tokio::select! {
        biased;
        _ = ticker.tick() => (),
        _ = shutdown.notified() => break,
    }
}
```